### PR TITLE
Add simple KINETIC block parsing

### PR DESCRIPTION
**Pull request by halfflat**
Targets issue #62.

New tokens added:
* keywords `tok::kinetic`, `tok::conserve`;
* symbols `tok::tilde`, `tok::arrow`;
* split `tok::number` into `tok::real` and `tok::integer` (stoichiometric expressions may have only integer coefficients).

New expression sub-types added, together with parsing:
* `IntegerExpression` (derives from `NumberExpression`);
* `ReactionExpression` — represents a line of the form "~ a + b <-> c (f, r)";
* `StoichTermExpression` — represents a single term in the lhs or rhs of a reaction as an integral coefficient and identifier, for example "3c" or "-2 sp2";
* `StoichExpression` — represents a possibly empty series of `StoichTermExpression`s, corresponding to the lhs or rhs of a reaction, or the lhs of a `CONSERVE` declaration;
* `ConserveExpression` (derives from `BinaryExpression`) — represents a line of the form "CONSERVE a + 2b - c = expr".

Unit tests added for parsing and lexing of the above, with some clean up of the testing code:
* verbose testing selectable at run time with `-v`, instead of being a compile time decision;
* code duplication reduced in `test_parser.cpp` through the use of helper functions `check_parse` and `check_parse_fail`;
* handle verbose output from lexer tests by subclassing the lexer; additionally, emit tokens from `Lexer::parse()` in verbose output.

### CODE REVIEW

**Comment by halfflat** [comment]
Commenting just to test slack trigger (on master).

File: tests/modcc/test_lexer.cpp [hunk 1]
```diff
@@ -1,20 +1,45 @@
...
+    std::cout << "________________\n" << string << "\n________________\n";
+}
+
+void verbose_print(const Token& token) {
+    if (!g_verbose_flag) return;
+    std::cout << "tok: " << token << "\n";
+}
+
+class VerboseLexer: public Lexer {
+public:
```
Comment (bcumming) [inline tests/modcc/test_lexer.cpp]: I like the verbose lexer... very useful for debugging.

File: tests/modcc/test_parser.cpp [hunk 2]
```diff
@@ -15,481 +79,413 @@ TEST(Parser, full_file) {
...
     }
+}
 
-    {
-        Parser p("LOCAL x, ");
-        auto e = p.parse_local();
+// test parsing of line expressions
+TEST(Parser, parse_line_expression) {
+    const char* good_expr[] = {
+        "qt=q10^((celsius-22)/10)"
```
Comment (bcumming) [inline tests/modcc/test_parser.cpp]: Is this line missing a comma at the end?

File: tests/modcc/test_parser.cpp [hunk 3]
```diff
@@ -15,481 +79,413 @@ TEST(Parser, full_file) {
...
-    {
-"+x             ",
-"-x             ",
-"(x + -y)       ",
-"-(x - + -y)    ",
-"exp(x + y)     ",
-"-exp(x + -y)   ",
+TEST(Parser, parse_stoich_term) {
+    const char* good_pos_expr[] = {
+        "B", "B3", "3B3", "0A", "12A"
```
Comment (bcumming) [inline tests/modcc/test_parser.cpp]: what about `2e4` or `1E3` for example? These are ambiguous, if we have a variable named `e4` or `E3` respectively. From looking at the lexer and parser code, I think these might be a problem.

File: tests/modcc/test_parser.cpp [hunk 4]
```diff
@@ -15,481 +79,413 @@ TEST(Parser, full_file) {
...
-"foo(x+3, y, bar(21.4))",
-"y=exp(x+3) + log(exp(x/y))",
-"a=x^y^z",
-"a=x/y/z"
+// test parsing of stoich and reaction expressions
+TEST(Parser, parse_reaction_expression) {
+    const char* good_expr[] = {
+        "~ A + B <-> C + D (k1, k2)",
+        "~ 2B <-> C + D + E (k1(3,v), k2)",
+        "~ <-> C + D + 7 E (k1, f(a,b)-2)",
```
Comment (bcumming) [inline tests/modcc/test_parser.cpp]: Here you have a space between the number and the `E`. We discussed this didn't we, and decided that the compiler error message about an unexpected real number would be enough, didn't we?

File: tests/modcc/test_parser.cpp [hunk 5]
```diff
@@ -15,481 +79,413 @@ TEST(Parser, full_file) {
...
-#ifdef VERBOSE_TEST
-        if(e) std::cout << e->to_string() << std::endl;
-#endif
-        EXPECT_NE(e, nullptr);
-        EXPECT_EQ(p.status(), lexerStatus::happy);
+    const char* bad_expr[] = {
+        "~ A + B <-> C + D (k1, k2, k3)",
+        "~ A + B <-> C + (k1, k2)",
+        "~ 2.3B <-> C + D + E (k1(3,v), k2)",
+        "~ <-> C + D + 7E",
```
Comment (bcumming) [inline tests/modcc/test_parser.cpp]: Aha, we did decide that after all...

File: tests/modcc/test_parser.cpp (hunk 2 above)
Comment (halfflat) [inline tests/modcc/test_parser.cpp]: Indeed, yes!

**Pull request merged**